Added changes guided mode posvel controller to incorporate altitude controller

jverbeke opened the pull request:
In original pos-vel controller of guided mode, the position controller doesn't follow vertical velocity commands companion computer and doesn't compensate for terrein slope when using height sonar.
Changes are:

    Every time new position and velocity target message arrives, the new position is set with pos_control.set_pos_target(position). This function sets _flags.use_desvel_ff_z to false each time and sets _vel_desired.z to zero. These latter two are commented.

    Similar to pos controller in guided mode and loiter, but not present in the pos_and_vel controller, the alt target should be compensated by feedforward for the desired climb rate right before calling update_z_controller() and take into account the sonar height to cope with sloped terrains.

Cheers,

Jon

P.S. separate commits now

amilcarlucas commented on ArduCopter/control_guided.cpp:
```diff
@@ -137,6 +137,10 @@ void Copter::guided_posvel_control_start()
     pos_control->set_xy_target(curr_pos.x, curr_pos.y);
```
Please change the commit message of this file to:
Copter: Added altitude controller to guided posvel controller

including sonar alt adjustment

jverbeke replied:
done

amilcarlucas commented on libraries/AC_AttitudeControl/AC_PosControl.cpp:
```diff
@@ -531,8 +531,8 @@ void AC_PosControl::set_pos_target(const Vector3f& position)
 
-    _flags.use_desvel_ff_z = false;
-    _vel_desired.z = 0.0f;
+    //_flags.use_desvel_ff_z = false;
```
Please change this comit message to:
AC_AttitudeControl: Added altitude velocity feedforward for posvel controller

jverbeke replied:
done

lthall replied:
Why did you do this?

jverbeke replied:
Because when you use the combined position-velocity controller, the velocity commands didn’t go through.
The drone just kept hanging still in the air, regardless of any x, y or z velocity input.
This way the velocity input is directly fed forward.

lthall replied:
Did you consider what this would do to all the other modes that used this function?

amilcarlucas commented on ArduCopter/control_guided.cpp:
```diff
@@ -575,6 +579,16 @@ void Copter::guided_posvel_control_run()
+        // if sonar is ok, use surface tracking
+        float temp_surface_tracking_climbrate = get_surface_tracking_climb_rate((int)posvel_vel_target_cms.z, posvel_pos_target_cm.z, G_Dt);
+        pos_control.set_alt_target_from_climb_rate_ff(temp_surface_tracking_climbrate, G_Dt, false);
+    }
```
please change to
} else {

jverbeke replied:
Done

rmackay9 commented:
Thanks for the contribution.  You definitely had to dig into the details to figure out this change so hats off to you for that.

I think the intention here is to add terrain following to Guided mode's posvel controller which is great but I have a few concerns:

- the caller can't specify if the alt is a terrain alt or not.  This is possible with Auto mode thanks to the "frame" field in MISSION_ITEM (http://mavlink.org/messages/common#MISSION_ITEM).  It's also possible to choose in RTL using the poorly named TERRAIN_USE parameter.  In Loiter and other more manual flight modes the user has a switch to turn on/off use of the range finder with a switch.  So I think we need to use the SET_POSITION_TARGET_LOCAL_NED messages's "coordinate_frame" field to specify if the altitude is an alt-above-terrain or something else.
- the change only supports using a sonar/lidar, not the google earth terrain data
- the change only applies to the posvel controller while the other guided mode controls stay using non-terrain altitude.

jverbeke commented:
Hi @rmackay9 

I've added the sonar alt adjustment to the pos and vel controllers in guided. 
For the pure pos controller: I specified climb rate as zero
`get_surface_tracking_climb_rate(0, pos_control->get_alt_target(), G_Dt);`

I've also added the avoidance adjusted climb rate (from loiter mode) to all controllers. I assume this I what you meant by 

> the change only supports using a sonar/lidar, not the google earth terrain data

I've also updated some minor code syntax, which has changed since the 3.4 version I originally developed the posvel changes in.

I'm not completely following you with regards to the altitude being with respect to the terrain or not. If you sent a position command through the SET_POSITION_TARGET_LOCAL_NED, it gets added to the current position, also for altitude. The altitude within the current position, is GPS based and with respect to home altitude if I'm not mistaken. So I don't see why the alt being terrain or not is relevant. Could you clarify your comment further?

amilcarlucas commented:
Please rebase this on master

jverbeke commented:
Because when you use the combined position-velocity controller, the velocity commands didn’t go through.
The drone just kept hanging still in the air, regardless of any x, y or z velocity input.
This way the velocity input is directly fed forward.

Van: Leonard Hall [mailto:notifications@github.com]
Verzonden: maandag 15 januari 2018 0:47
Aan: ArduPilot/ardupilot <ardupilot@noreply.github.com>
CC: Jon Verbeke <jon.verbeke@kuleuven.be>; Author <author@noreply.github.com>
Onderwerp: Re: [ArduPilot/ardupilot] Added changes guided mode posvel controller to incorporate altitude controller (#7211)


@lthall commented on this pull request.

________________________________

In libraries/AC_AttitudeControl/AC_PosControl.cpp<https://github.com/ArduPilot/ardupilot/pull/7211#discussion_r161414619>:

> @@ -531,8 +531,8 @@ void AC_PosControl::set_pos_target(const Vector3f& position)

 {

     _pos_target = position;



-    _flags.use_desvel_ff_z = false;

-    _vel_desired.z = 0.0f;

+    //_flags.use_desvel_ff_z = false;

Why did you do this?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/ArduPilot/ardupilot/pull/7211#discussion_r161414619>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AHD9b7wp_MYB1Stofcbr7JX7Fw6bt4NZks5tKpIKgaJpZM4QX6b6>.
{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/ArduPilot/ardupilot","title":"ArduPilot/ardupilot","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/ArduPilot/ardupilot"}},"updates":{"snippets":[{"icon":"PERSON","message":"@lthall commented on #7211"}],"action":{"name":"View Pull Request","url":"https://github.com/ArduPilot/ardupilot/pull/7211#discussion_r161414619"}}}

amilcarlucas commented:
@rmackay9 , @jverbeke needs some clarifications.

peterbarker commented:
We've had massive changes since this one in, including a massive rewrite of position control.

Sadly, I think this has bitrotted too far.

Apologies we couldn't conclude this in a better way, @jverbeke - but I think I'll close this now.

If you do want to chase it, please feel free to open a new PR.

amilcarlucas commented:
fixes #6736
